src/python/gennylib/auto_tasks.py

```diff
@@ -0,0 +1,361 @@
+
+
+class Runtime:
+    use_expansions_yml: bool = False
+
+    def __init__(self, cwd: str, conts: Optional[dict] = None):
+        self.conts = conts
+        self.cwd = cwd
+
+    def _load(self):
+        if self.conts is not None:
+            return
+        conts = _yaml_load(
+            [os.path.join(self.cwd, f"{b}.yml") for b in {"bootstrap", "runtime", "expansions"}]
+        )
+        if "expansions" in conts:
+            self.use_expansions_yml = True
+            conts = conts["expansions"]
+        else:
+            if "bootstrap" not in conts:
+                raise Exception(
+                    f"Must have either expansions.yml or bootstrap.yml in cwd={self.cwd}"
+                )
+            bootstrap = conts["bootstrap"]  # type: dict
+            runtime = conts["runtime"]  # type: dict
+            bootstrap.update(runtime)
+            conts = bootstrap
+        self.conts = conts
+
+    def has(self, key: str, acceptable_values: List[str]) -> bool:
+        self._load()
+        if key not in self.conts:
+            raise Exception(f"Unknown key {key}. Know about {self.conts.keys()}")
+        actual = self.conts[key]
+        return any(actual == acceptable_value for acceptable_value in acceptable_values)
+
+
+class GeneratedTask(NamedTuple):
+    name: str
+    mongodb_setup: Optional[str]
+    workload: "Workload"
+
+
+class Workload:
+    file_path: str
+    is_modified: bool
+    requires: Optional[dict] = None
+    setups: Optional[List[str]] = None
+
+    def __init__(self, file_path: str, is_modified: bool, conts: Optional[dict] = None):
+        self.file_path = file_path
+        self.is_modified = is_modified
+
+        if not conts:
+            with open(file_path) as conts:
+                conts = yaml.safe_load(conts)
+
+        if "AutoRun" not in conts:
+            return
+
+        auto_run = conts["AutoRun"]
+        self.requires = auto_run["Requires"]
+        if "PrepareEnvironmentWith" in auto_run:
+            prep = auto_run["PrepareEnvironmentWith"]
+            if len(prep) != 1 or "mongodb_setup" not in prep:
+                raise ValueError(
+                    f"Need exactly mongodb_setup: [list] "
+                    f"in PrepareEnvironmentWith for file {file_path}"
+                )
+            self.setups = prep["mongodb_setup"]
+
+    def file_base_name(self) -> str:
+        return str(os.path.basename(self.file_path).split(".yml")[0])
+
+    def relative_path(self) -> str:
+        return self.file_path.split("src/workloads/")[1]
+
+    def all_tasks(self) -> List[GeneratedTask]:
+        base = self._to_snake_case(self.file_base_name())
+        if self.setups is None:
+            return [GeneratedTask(base, None, self)]
+        return [
+            GeneratedTask(f"{base}_{self._to_snake_case(setup)}", setup, self)
+            for setup in self.setups
+        ]
+
+    def variant_tasks(self, runtime: Runtime) -> List[GeneratedTask]:
```

Comment:
I didn't quite get how the 3 `variant_tasks` function fit together when skimming this file. Asking bc I'm curious how we can use this to get to the cheap/expensive buildvariant world you described the other day.

Reply:
One is on the `Workload` that indicates what tasks (if any) the workload yaml file contributes to the tasks for the given variant. Another is on the `Repo` which basically flatmaps over all its workloads `variant_tasks` values. The final is on the writer which translates the Repo's value to the shrub config.

It seemed simpler to me in code than in writing it just now, so I'll ensure this is clear by names or comments. Good thing to point out.

> how we can use this to get to the cheap/expensive buildvariant world

I'm not sure to be honest - not sure how we list the various combinations of bootstrap inputs if we don't do it by build-variant. Maybe each workload lists the runs it wants to have:

```yaml
AutoRun:
  Expensive:
  - mongodb_setup: replica
    infrastructure_provisioning: replica
```

but that may be too much tying Genny workloads to DSI stuff.

Thankfully I think I've got the right level of factoring in here to allow such a change to be contained to just the `Workload` class 😸

Reply:
SGTM